Mathlib/LinearAlgebra/PiTensorProduct.lean

```diff
@@ -463,17 +472,104 @@ then there is an induced map `⨂ᵢ sᵢ → ⨂ᵢ tᵢ` by `⨂ aᵢ ↦ ⨂
+def map : (⨂[R] i, s i) →ₗ[R] ⨂[R] i, t i :=
   lift <| (tprod R).compLinearMap f
 
-@[simp] lemma map_tprod (f : Π i, s i →ₗ[R] t i) (x : Π i, s i) :
+@[simp] lemma map_tprod (x : Π i, s i) :
     map f (tprod R x) = tprod R fun i ↦ f i (x i) :=
   lift.tprod _
 
+-- No lemmas about associativity, because we don't have associativity of `PiTensorProduct` yet.
+
+theorem map_range_eq_span_tprod :
+    LinearMap.range (map f) =
+      Submodule.span R { t | ∃ (m : (i : ι) → s i), tprod R (fun i ↦ f i (m i)) = t } := by
+  rw [← Submodule.map_top, ← span_tprod_eq_top, Submodule.map_span, ← Set.range_comp]
+  apply congrArg; ext x
+  simp only [Set.mem_range, comp_apply, map_tprod, Set.mem_setOf_eq]
+
+/-- Given submodules `p i ⊆ s i`, this is the natural map: `⨂[R] i, p i → ⨂[R] i, s i`.
+This is `TensorProduct.mapIncl` for an arbitrary family of modules.
+-/
+@[simp]
+def mapIncl (p : (i : ι) → Submodule R (s i)) : (⨂[R] i, p i) →ₗ[R] ⨂[R] i, s i :=
+  map (fun (i : ι) ↦ (p i).subtype)
+
+theorem map_comp : map (fun (i : ι) ↦ g i ∘ₗ f i) = (map g) ∘ₗ (map f) := by
+  ext
+  simp only [LinearMap.compMultilinearMap_apply, map_tprod, LinearMap.coe_comp, Function.comp_apply]
+
+theorem lift_comp_map (h : MultilinearMap R t E) :
+    lift h ∘ₗ map f = lift (h.compLinearMap f) := by
+  ext
+  simp only [LinearMap.compMultilinearMap_apply, LinearMap.coe_comp, Function.comp_apply,
+    map_tprod, lift.tprod, MultilinearMap.compLinearMap_apply]
+
+attribute [local ext high] ext
+
+@[simp]
+theorem map_id : map (fun i ↦ (LinearMap.id : s i →ₗ[R] s i))  = .id := by
+  ext
+  simp only [LinearMap.compMultilinearMap_apply, map_tprod, LinearMap.id_coe, id_eq]
+
+@[simp]
+theorem map_one : map (fun (i : ι) ↦ (1 : s i →ₗ[R] s i)) = 1 :=
+  map_id
+
+theorem map_mul (f₁ f₂ : Π i, s i →ₗ[R] s i)  :
+    map (fun i ↦ f₁ i * f₂ i) = map f₁ * map f₂ :=
+  map_comp f₁ f₂
+
+/-- Upgrading `PiTensorProduct.map` to a `MonoidHom` when `s = t`.-/
+def mapMonoidHom : (Π i, s i →ₗ[R] s i) →* ((⨂[R] i, s i) →ₗ[R] ⨂[R] i, s i) where
+  toFun := map
+  map_one' := map_one
+  map_mul' := map_mul
+
+@[simp]
+protected theorem map_pow (f : Π i, s i →ₗ[R] s i) (n : ℕ) :
+    map (f ^ n) = map f ^ n := MonoidHom.map_pow mapMonoidHom _ _
+
+open Function in
+theorem map_add_smul_aux [DecidableEq ι] (i : ι) (x : Π i, s i) (u : s i →ₗ[R] t i) :
+    (fun (j : ι) ↦ (update f i u j) (x j)) = update (fun (j : ι) ↦ (f j) (x j)) i (u (x i)) := by
+  ext j
+  by_cases h : j = i
+  · rw [h]; simp only [update_same]
+  · simp only [ne_eq, h, not_false_eq_true, update_noteq]
+
+open Function in
+theorem map_add [DecidableEq ι] (i : ι) (u v : s i →ₗ[R] t i) :
+    map (update f i (u + v)) = map (update f i u) + map (update f i v) := by
+  ext x
+  simp only [LinearMap.compMultilinearMap_apply, map_tprod, map_add_smul_aux, LinearMap.add_apply,
+    MultilinearMap.map_add]
+
+open Function in
+theorem map_smul [DecidableEq ι] (i : ι) (c : R) (u : s i →ₗ[R] t i) :
```

Comment:
I think these should be `map_update_add` etc

Reply:
Here's my logic behind the naming: `PiTensorProduct.map` is a `MultilinearMap` (in fact, we are in the process of proving it), and these lemmas `map_add` and `map_smul` are proving the same things as `MultilinearMap.map_add` and `MultilinearMap.map_smul`. So they should have the same names.

Reply:
I guess I don't care that much, so am happy to go with that argument. I agree it's consistent.